### PR TITLE
Add new command fw_meta_info for getting metadata info from Flash partition

### PR DESCRIPTION
**Pull request by rakeshvitta**
This change is to add a new command "fw_meta_info" to get the Metadata information from the flash partitions. 

Changes include:
1. New function fw_meta_info() to handle the command. Metadata info is derived by calling the switchtec_fw_part_info() function that already exists in fw.c which gets the required info from the switch.
2. Helper function part_to_string() to convert partition ID to partition name string.
3. Add the definition of switchtec_fw_part_info() to switchtec.h file instead of it being a static function in fw.c.
4. Moved the structure switchtec_fw_metadata_gen4 from fw.c to switchtec.h

Thanks and Regards,
Rakesh Vitta

### CODE REVIEW

File: cli/main.c [hunk 1]
```diff
@@ -1434,6 +1434,88 @@ static int fw_info(int argc, char **argv)
 	return 0;
 }
 
+static char * part_to_string(int partition_id) {
+       switch(partition_id){
+
```
Comment (lsgunth) [inline cli/main.c:1439]: We use the Linux code style in this project. Please review the documentation for that before submitting:

https://www.kernel.org/doc/html/latest/process/coding-style.html

The first two lines have a ton of style issues:

- Pointer should not have a space after the asterisk
- The open brace for a function should be on it's own line
- There should be a space after "switch"
- There should be a space before the open brace of switch
- Tabs should be used instead of 8 spaces.

File: inc/switchtec/switchtec.h [hunk 2]
```diff
@@ -241,6 +241,38 @@ struct switchtec_fw_image_info {
 	unsigned long secure_version;
 };
 
+struct switchtec_fw_metadata_gen4 {
```
Comment (lsgunth) [inline inc/switchtec/switchtec.h:244]: Changes to the library should be in another patch.

This is a low level hardware interface that should not be exposed outside of the library.

If we need to print more information we should extend the existing interfaces, not bypass them.

File: cli/main.c [hunk 3]
```diff
@@ -1434,6 +1434,88 @@ static int fw_info(int argc, char **argv)
...
+               int partition_id;
+       } cfg = {};
+
+       struct switchtec_fw_image_info inf;
+       struct switchtec_fw_metadata_gen4 *metadata;
+
+       const struct argconfig_options opts[] = {
+               DEVICE_OPTION,
+               {"flash_part", 'p', "", CFG_NONNEGATIVE, &cfg.partition_id, required_argument,
+                       "Flash Partition ID"},
```
Comment (lsgunth) [inline cli/main.c:1470]: How does the user know which number for which partition ID? Why should the CLI be creating a new number for this? Other CLI commands use short strings to reference the partition type ("BOOT", "MAP", etc) not an opaque number.

**Review by lsgunth — CHANGES_REQUESTED** [review]
It's not clear what the benefit here is. Please describe in the commit message what this new command is providing over the existing firmware info interfaces.

**Comment by rakeshvitta** [comment]
> It's not clear what the benefit here is. Please describe in the commit message what this new command is providing over the existing firmware info interfaces.

This command prints out the metadata information for the flash partition specified. The metadata includes information like public key modulus, exponent and secure version which are not shown in the fw-info command.

File: cli/main.c (hunk 1 above)
Comment (rakeshvitta) [reply cli/main.c]: Fixed the style issues and checked the patch using checkpatch.pl script. 
Commit: 10889e6acb50ac44dce8a53e9cd622d993713aba

File: inc/switchtec/switchtec.h (hunk 2 above)
Comment (rakeshvitta) [reply inc/switchtec/switchtec.h]: Done. Implemented the library changes as a different commit. I have put the metadata definition back to the library and added new variables to the existing structure to expose the metadata fields from the flash partition. 
I had to expose the switchtec_part_info() as this provides a interface to get the flash info for a particular partition instead of all the flash partitions that the current exposed function provides.

File: cli/main.c (hunk 3 above)
Comment (rakeshvitta) [reply cli/main.c]: This partition ID is the partition number in the flash and device spec has the details. We cannot use the short strings as they only specify what type of partition it is and does not specify if we are looking at the active or redundant type of the partition.

Comment (kelvin-cao) [reply cli/main.c]: The partition ID can be inconsistent over generations in the firmware. Using short strings provides better compatibility.

**Comment by kelvin-cao** [comment]
> 
> 
> > It's not clear what the benefit here is. Please describe in the commit message what this new command is providing over the existing firmware info interfaces.
> 
> This command prints out the metadata information for the flash partition specified. The metadata includes information like public key modulus, exponent and secure version which are not shown in the fw-info command.

Current fw-info provides some info from meta data, like CRC. We don't need a brand new command just provide the info that the meta data MRPC command returns. Instead, maybe we can introduce a '-v | --verbose' mode to the fw-info to include these extra data. (We don't want to change the default fw-info behavior as it will hurt compatibility)

**Comment by kelvin-cao** [comment]
> 
> 
> > It's not clear what the benefit here is. Please describe in the commit message what this new command is providing over the existing firmware info interfaces.
> 
> This command prints out the metadata information for the flash partition specified. The metadata includes information like public key modulus, exponent and secure version which are not shown in the fw-info command.

Current fw-info provides some info from meta data, like CRC. We don't need a brand new command just provide the info that the meta data MRPC command returns. Instead, maybe we can introduce a '-v | --verbose' mode to the fw-info to include these extra data. (We don't want to change the default fw-info behavior as it will hurt compatibility)

**Comment by kelvin-cao** [comment]
To fix issues in a patch in the PR, you may want to re-base to the patch and fix the issues then hard push them, instead of adding a new patch. And there should be no CI check failures.

**Comment by rakeshvitta** [comment]
Will close this pull request and add the code under fw-info with a --v option instead of adding a brand new command to print the meta-data info of a partition. Since the code would be very different from what I added here will close this pull request and create a new one.
